qiskit_experiments/data_processing/nodes.py

```diff
@@ -411,6 +433,135 @@ def _process(self, data: np.array) -> np.array:
+        self._no_ufloat = no_ufloat
+
+    def _format_data(self, data: np.ndarray) -> np.ndarray:
+        """Check that there are as many discriminators as there are slots."""
+        data = super()._format_data(data)
+
+        if self._validate:
+            if isinstance(self._discriminator, list):
+                if self._n_slots != len(self._discriminator):
+                    raise DataProcessorError(
+                        f"The Discriminator node has {len(self._discriminator)} which does "
+                        f"not match the {self._n_slots} slots in the data."
+                    )
+
+        if self._no_ufloat:
+            data = unp.nominal_values(data)
+
+        return data
+
+    def _process(self, data: np.ndarray) -> np.ndarray:
+        """Discriminate the data.
+
+        Args:
+            data: The IQ data as a list of points to discriminate. This data should have
+                the shape dim_1 x dim_2 x ... x dim_k x 2.
+
+        Returns:
+            The discriminated data as a list of labels with shape dim_1 x ... x dim_k.
+        """
+        # Case where one discriminator is applied to all the data.
+        if not isinstance(self._discriminator, list):
+            # Reshape the IQ data to an array of size n x 2
+            shape, data_length = data.shape, 1
+            for dim in shape[0:-1]:
+                data_length *= dim
+
+            data = data.reshape((data_length, 2))  # the last dim is guaranteed by _process
+
+            # Classify the data using the discriminator and reshape it to dim_1 x ... x dim_k
+            classified = np.array(self._discriminator.predict(data)).reshape(shape[0:-1])
+
+        # case where a discriminator is applied to each slot.
+        else:
+            if self._n_shots == 0:
+                classified = np.empty((self._n_circs, self._n_slots), dtype=str)
+                for idx, discriminator in enumerate(self._discriminator):
+                    sub_data = data[:, idx, :].reshape((self._n_circs, 2))
+                    sub_classified = discriminator.predict(sub_data)
+                    classified[:, idx] = sub_classified
+
+            else:
+                classified = np.empty((self._n_circs, self._n_shots, self._n_slots), dtype=str)
+                for idx, discriminator in enumerate(self._discriminator):
+                    sub_data = data[:, :, idx, :].reshape((self._n_circs * self._n_shots, 2))
+                    sub_classified = np.array(discriminator.predict(sub_data))
+                    sub_classified = sub_classified.reshape((self._n_circs, self._n_shots))
+                    classified[:, :, idx] = sub_classified
+
+        # Concatenate the bit-strings together.
+        # Averaged data
+        if self._n_shots == 0:
+            # ::-1 for Qiskit's convention, e.g. in "001" qubit 0 is in state "1".
+            labeled_data = ["".join(classified[idx, :][::-1]) for idx in range(self._n_circs)]
```

Comment:
Why does it require the labels of the discriminator to be strings instead of ints? sklearn classifiers produce ints so I would think that's more natural.